### PR TITLE
add checkpoint util class and implement

### PR DESCRIPTION
**Pull request by seiriosPlus**
Add new feature about https://github.com/PaddlePaddle/Paddle/issues/10376

### CODE REVIEW

**Comment by seiriosPlus** [comment]
目前还有一些细节点需要优化..改进中

File: paddle/fluid/operators/checkpoint_load_op.cc [hunk 1]
```diff
@@ -0,0 +1,214 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (typhoonzero) [inline paddle/fluid/operators/checkpoint_load_op.cc]: I see there's `load_op` and `load_combine_op` and corresponding saving ops, on python side, you can also use `fluid.io.save_persistables` to save all persistable variables.

In order to make `save_persistables` equal to save a checkpoint, make sure that the state variables are all "persistable" like step counters, learning rates, learning_rate moments etc.

So can you reuse those ops instead of writing some new one?

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 2]
```diff
@@ -187,6 +187,7 @@ void ListenAndServOp::RunSyncLoop(framework::Executor *executor,
     for (auto &var : sparse_vars) {
       var->GetMutable<framework::SelectedRows>()->mutable_rows()->clear();
     }
+
```
Comment (typhoonzero) [inline paddle/fluid/operators/listen_and_serv_op.cc]: This change may not be nessesary.

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 3]
```diff
@@ -315,6 +326,24 @@ def transpile(self,
                 "epmap": eplist,
                 "sync_mode": self.sync_mode
             })
+
+        if checkpoint_dir and self.is_chief:
+            program.global_block().create_var(
+                name=SERIAL_VAR_NAME,
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: what is a "serial"?

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 4]
```diff
@@ -315,6 +326,24 @@ def transpile(self,
...
+                persistable=True,
+                type=core.VarDesc.VarType.RAW)
+
+            save_vars = []
+            for var in self.origin_program.list_vars():
+                if self._is_persistable(var):
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: can use `fluid.io.save_persistables`

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 5]
```diff
@@ -547,8 +605,61 @@ def _get_splited_name_and_shape(varname):
...
                     attrs=op.attrs)
+                for var in new_outputs.values():
+                    load_vars.append(var.name)
+        # add checkpoint op
+        if not checkpoint_load_dir:
+            return s_prog
+
+        serial_number = self._get_lastest_checkpoint_dir(checkpoint_load_dir)
+
+        s_prog.global_block().append_op(
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: How do current parameter server know which parameter block to load?

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 6]
```diff
@@ -210,6 +216,11 @@ def transpile(self,
         self.pserver_endpoints = pserver_endpoints
         self.optimize_ops, params_grads = self._get_optimize_pass()
 
+        # is_chief (no.0 triner) for checkpoint
+        # the no.0 trainer will save all variables and its own reader offset to checkpoint
+        # other trianers will save its own reader offset to checkpoint
+        self.is_chief = trainer_id == 0
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: _is_chief

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 7]
```diff
@@ -547,8 +605,61 @@ def _get_splited_name_and_shape(varname):
...
+            """
+            if not os.path.isdir(cur_dir):
+                return -1
+
+            try:
+                int(cur_dir)
+            except ValueError:
+                return -1
+
+            success_path = os.path.join(checkpoint_dir, cur_dir, SUCCESS)
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: what is success_path used for?

Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: We need a tag to indicate that: the checkpoint content is right. So, I define a tag named ```_SUCCESS```, when the ```checkpoint_save_op``` save all need variables successfully, it will write an empty file named ```_SUCCESS``` at last.
Because of this, when pserver/trainer need to load ```checkpoint```, it will check ```_SUCCESS``` at first.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 6 above)
Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: I will fix it.

File: paddle/fluid/operators/listen_and_serv_op.cc (hunk 2 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/listen_and_serv_op.cc]: I am sorry about it, I will revert it later.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 7 above)
Comment (typhoonzero) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: If you only need to know when the saving ends, just wait for the `executor` returns, raise any error that may cause the saving fail is OK I think.

File: paddle/fluid/operators/checkpoint_load_op.cc (hunk 1 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/checkpoint_load_op.cc]: ```load_op``` and ```save_op``` are designed for ```LodTensor variable```,  But checkpoint will save variables not only ```LodTensor```， and ```checkpoint``` has some arguments particular.
At present, ```checkpoint load/save op``` and ```load/save op``` have no clear-cut distinction.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 7 above)
Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: If some exceptions happened in ```executor ``` running, the ```executor ``` does not have any information return, how do we know saving is OK.

Comment (typhoonzero) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: You can catch that exception I think, please give it a try, this will make the code more simple.

File: paddle/fluid/operators/checkpoint_load_op.cc (hunk 1 above)
Comment (typhoonzero) [reply paddle/fluid/operators/checkpoint_load_op.cc]: I think it's better to reuse current operators maybe check the variable type will be fine.
So what the other variable types are saved in the checkpoint? "RAW" types and "feed" "fetch" may not need to be saved.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 3 above)
Comment (typhoonzero) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: Call it a checkpoint ID maybe good for understanding

File: python/paddle/fluid/io.py [hunk 8]
```diff
@@ -454,3 +452,149 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+BEGIN_SECS = None
+
+
+def save_checkpoint(executor,
+                    dirname,
+                    keep_max=3,
+                    save_secs=600,
+                    main_program=None):
+    """
+    Save Variables to Checkpint Dir
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: Checkpint => Checkpoint
Dir => Directory

File: python/paddle/fluid/io.py [hunk 9]
```diff
@@ -454,3 +452,149 @@ def get_parameter_value_by_name(name, executor, program=None):
...
     return get_parameter_value(var, executor)
+
+
+SUCCESS = "_SUCCESS"
+BEGIN_SECS = None
+
+
+def save_checkpoint(executor,
+                    dirname,
+                    keep_max=3,
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: keep_max => max_num_checkpoints

File: python/paddle/fluid/io.py [hunk 10]
```diff
@@ -454,3 +452,149 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+                    save_secs=600,
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: save_secs => interval

File: python/paddle/fluid/io.py [hunk 11]
```diff
@@ -454,3 +452,149 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+
+    global BEGIN_SECS
+    if BEGIN_SECS is not None:
+        if time.time() - BEGIN_SECS < save_secs:
+            return
+    BEGIN_SECS = time.time()
+
+    serial = _get_lastest_checkpoint_dir(dirname) + 1
+    cur_dir = os.path.join(dirname, str(serial))
+    # save_persistables(executor, cur_dir, main_program)
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: No commented out codes please.

File: python/paddle/fluid/io.py [hunk 12]
```diff
@@ -454,3 +452,149 @@ def get_parameter_value_by_name(name, executor, program=None):
...
     var = program.global_block().var(name)
     return get_parameter_value(var, executor)
+
+
+SUCCESS = "_SUCCESS"
+BEGIN_SECS = None
+
+
+def save_checkpoint(executor,
+                    dirname,
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: Can use current working directory as default?

File: python/paddle/fluid/io.py [hunk 13]
```diff
@@ -454,3 +452,149 @@ def get_parameter_value_by_name(name, executor, program=None):
         program = default_main_program()
     var = program.global_block().var(name)
     return get_parameter_value(var, executor)
+
+
+SUCCESS = "_SUCCESS"
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: SUCCESS = SUCCESS_MARK_FILENAME

File: python/paddle/fluid/io.py [hunk 14]
```diff
@@ -454,3 +452,149 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+    :param save_secs
+    :param main_program
+    """
+    if dirname is None:
+        raise Exception("save checkpoint dir can not be none")
+
+    if not os.path.isdir(dirname):
+        os.makedirs(dirname)
+
+    global BEGIN_SECS
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: try not to use global please

File: python/paddle/fluid/io.py [hunk 15]
```diff
@@ -454,3 +452,149 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+    if BEGIN_SECS is not None:
+        if time.time() - BEGIN_SECS < save_secs:
+            return
+    BEGIN_SECS = time.time()
+
+    serial = _get_lastest_checkpoint_dir(dirname) + 1
+    cur_dir = os.path.join(dirname, str(serial))
+    # save_persistables(executor, cur_dir, main_program)
+    save_vars(
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: why call `save_vars` instead of `save_psersistables`?

Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: save_psersistables can not filter out gradient variables.

File: python/paddle/fluid/io.py (hunk 13 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py (hunk 12 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py (hunk 14 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py (hunk 11 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py (hunk 9 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py (hunk 10 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py (hunk 8 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py [hunk 16]
```diff
@@ -13,21 +13,17 @@
...
-    'save_persistables',
-    'load_vars',
-    'load_params',
-    'load_persistables',
-    'save_inference_model',
-    'load_inference_model',
-    'get_inference_program',
+    'save_vars', 'save_params', 'save_persistables', 'load_vars', 'load_params',
+    'load_persistables', 'save_inference_model', 'load_inference_model',
+    'get_inference_program', 'save_checkpoint', 'restore_checkpoint'
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: maybe it's better to be named as `load_checkpoint` or `restore_from_checkpoint`?

File: python/paddle/fluid/io.py [hunk 17]
```diff
@@ -454,3 +452,160 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+
+
+def _interval_secs_exceed(dirname, save_interval_secs):
+    dir_time = os.path.getmtime(dirname)
+    if save_interval_secs > (time.time() - dir_time):
+        return False
+    return True
+
+
+def _lru_delete(dirname, keep_max=3):
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: keep_max => max_num_checkpoints

File: python/paddle/fluid/io.py [hunk 18]
```diff
@@ -454,3 +452,160 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+    """
+    retain checkpoint nums with keep_max
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: keep_max => max_num_checkpoints

File: python/paddle/fluid/io.py [hunk 19]
```diff
@@ -454,3 +452,160 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+    serials.sort(reverse=True)
+    serials = serials[keep_max:]
+    for serial in serials:
+        cur_dir = os.path.join(dirname, str(serial))
+        shutil.rmtree(cur_dir)
+
+
+def _write_success(dirname):
+    """
+    write _SUCCESS to checkpoint dir
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: update `_SUCCESS`

File: python/paddle/fluid/io.py [hunk 20]
```diff
@@ -454,3 +452,160 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+    """
+    success_file = os.path.join(dirname, SUCCESS_MARK_FILENAME)
+    with open(success_file, 'a'):
+        pass
+
+
+def _get_lastest_checkpoint_dir(checkpoint_dir):
+    """
+    get the biggest number in checkpoint_dir, which has _SUCCESS
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: update _SUCCESS

File: python/paddle/fluid/io.py (hunk 16 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: I will use ```load_checkpoint ```, ```restore_from_checkpoint ``` is too long I think.

File: python/paddle/fluid/io.py (hunk 19 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: ```_SUCCESS ``` is the file name ,not the var name

File: python/paddle/fluid/io.py (hunk 20 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: ```_SUCCESS ``` is the file name, not the var name

File: python/paddle/fluid/io.py (hunk 17 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py (hunk 16 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py (hunk 18 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py [hunk 21]
```diff
@@ -454,3 +452,167 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+    if not os.path.isdir(dirname):
+        os.makedirs(dirname)
+
+    serial = _get_lastest_checkpoint_dir(dirname)
+    if serial >= 0 and not _interval_secs_exceed(
+            os.path.join(dirname, str(serial)), save_interval_secs):
+        return
+
+    serial = serial + 1
+    cur_dir = os.path.join(dirname, str(serial))
```
Comment (typhoonzero) [inline python/paddle/fluid/io.py]: The checkpoint directories will be saved in "1", "2", etc. which may not make sense to users, I think it's better to be like "checkpoint_1", "checkpoint_2", and the "_SUCCESS" file can save a timestamp when the checkpoint is saved.

File: python/paddle/fluid/io.py [hunk 22]
```diff
@@ -454,3 +452,167 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+
+    if not os.path.isdir(dirname):
+        os.makedirs(dirname)
+
+    serial = _get_lastest_checkpoint_dir(dirname)
+    if serial >= 0 and not _interval_secs_exceed(
+            os.path.join(dirname, str(serial)), save_interval_secs):
+        return
+
+    serial = serial + 1
```
[the remark this answers is not in the document]
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py [hunk 23]
```diff
@@ -454,3 +452,167 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+        vars=None,
+        predicate=_is_checkpoint_var,
+        filename=None)
+    _write_success(cur_dir)
+    _lru_delete(dirname, max_num_checkpoints)
+
+
+def load_checkpoint(executor, dirname=None, main_program=None):
+    """
+    Load checkpoint from directory by executor,
```
[the remark this answers is not in the document]
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/io.py [hunk 24]
```diff
@@ -454,3 +452,167 @@ def get_parameter_value_by_name(name, executor, program=None):
...
+def save_checkpoint(executor,
+                    dirname=None,
+                    max_num_checkpoints=3,
+                    save_interval_secs=600,
+                    main_program=None):
+    """
+    Save Checkpoint will save persistable LodTensor variables from main_program in checkpoint directory,
+    directory named by serial number from 0 to (n -1), save_checkpoint use LRU strategy
+    to keep numbers of checkpoint directory,  the numbers of checkpoint directory are max_num_checkpoints at most,
+    The interval time between two save_checkpoint must great than or equal to save_interval_secs.
```
[the remark this answers is not in the document]
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done
